pkg/backend/filestate/meta.go

```diff
@@ -66,13 +73,40 @@ func ensurePulumiMeta(ctx context.Context, b Bucket) (*pulumiMeta, error) {
 		return meta, nil
 	}
 
-	// If there's no metadata file, we need to create one
-	// with the latest version.
+	// If there's no metadata file, we need to create one.
+	// The version we pick for the new file decides how we lay out the state.
+	//
+	// - Version 0 is legacy mode, which is the old layout.
+	//   To avoid breaking old stacks, we want to use version 0
+	//   if the bucket is not empty.
+	//
+	// - Version 1 added support for project-scoped stacks.
+	//   For entirely new buckets, we'll use version 1
+	//   to give new users access to the latest features.
 	//
 	// Implementation detail:
-	// For version 0, we don't write the file.
-	// However, for future versions, we will write it.
-	meta = &pulumiMeta{Version: 0}
+	// For version 0, we don't write the meta.yaml file.
+	empty, err := isPulumiDirEmpty(ctx, b)
+	if err != nil {
+		return nil, err
+	}
+
+	useLegacy := !empty
+	if empty {
+		// Allow opting into legacy mode for new states
+		// by setting the environment variable.
+		v, err := strconv.ParseBool(os.Getenv(PulumiFilestateLegacyLayoutEnvVar))
+		if err == nil {
+			useLegacy = v
+		}
+	}
+
+	if useLegacy {
+		meta = &pulumiMeta{Version: 0}
```

Comment:
Shouldn't there be a `return meta, nil` here to skip the write?

Reply:
Ah I see WriteTo does the skip, a little bit hidden maybe that should be pointed out in the comment above?